### PR TITLE
Add from_timestamp_micros function

### LINKED ISSUE #904: Feature request: `NaiveDateTime::from_timestamp_micros`
There is already `NaiveDateTime::from_timestamp_millis`, which uses some non-trivial logic to convert properly to seconds and nanoseconds when dealing with pre-1970 timestamps.

When instantiating one from a timestamp in _microseconds_ (which we have seen in the wild from some of our customers), it's necessary to duplicate that logic; see e.g. https://github.com/MaterializeInc/materialize/pull/16609 .

It would be really nice if Chrono had a `from_timestamp_micros` feature for this case.

### PR DESCRIPTION
**Pull request by umanwizard**
Fixes #904

### CODE REVIEW

File: src/naive/datetime/tests.rs [hunk 1]
```diff
@@ -31,6 +31,43 @@ fn test_datetime_from_timestamp_millis() {
...
+        (1662921288123456, "2022-09-11 18:34:48.123456000"),
+        (1662921287890000, "2022-09-11 18:34:47.890000000"),
+        (-2208936075000000, "1900-01-01 14:38:45.000000000"),
+        (0, "1970-01-01 00:00:00.000000000"),
+        (119731017000000, "1973-10-17 18:36:57.000000000"),
+        (1234567890000000, "2009-02-13 23:31:30.000000000"),
+        (2034061609000000, "2034-06-16 09:06:49.000000000"),
+    ];
+
+    for (timestamp_micros, formatted) in valid_map.iter().cloned() {
```
Comment (djc) [inline src/naive/datetime/tests.rs:47]: Nit: `cloned()` -> `copied()`.

File: src/naive/datetime/tests.rs [hunk 2]
```diff
@@ -31,6 +31,43 @@ fn test_datetime_from_timestamp_millis() {
...
+        let naive_datetime = NaiveDateTime::from_timestamp_micros(timestamp_micros);
+        assert_eq!(timestamp_micros, naive_datetime.unwrap().timestamp_micros());
+        assert_eq!(naive_datetime.unwrap().format("%F %T%.9f").to_string(), formatted);
+    }
+
+    let invalid = [i64::MAX, i64::MIN];
+
+    for timestamp_micros in invalid.iter().cloned() {
```
Comment (djc) [inline src/naive/datetime/tests.rs:55]: Same here.

File: src/naive/datetime/tests.rs [hunk 3]
```diff
@@ -31,6 +31,43 @@ fn test_datetime_from_timestamp_millis() {
...
+        let naive_datetime = NaiveDateTime::from_timestamp_micros(timestamp_micros);
+        assert!(naive_datetime.is_none());
+    }
+
+    // Test that the result of `from_timestamp_micros` compares equal to
+    // that of `from_timestamp_opt`.
+    let secs_test = [0, 1, 2, 1000, 1234, 12345678, -1, -2, -1000, -12345678];
+    for secs in secs_test.iter().cloned() {
```
Comment (djc) [inline src/naive/datetime/tests.rs:63]: And here.

File: src/naive/datetime/mod.rs [hunk 4]
```diff
@@ -129,6 +155,36 @@ impl NaiveDateTime {
         datetime.expect("invalid or out-of-range datetime")
     }
 
+    /// Creates a new [NaiveDateTime] from milliseconds or microseconds since the UNIX epoch.
```
Comment (djc) [inline src/naive/datetime/mod.rs:158]: Style nit: I would prefer this (private) API to be ordered below the public users of this method.

**Review by djc — APPROVED** [review]
LGTM, thanks!

**Comment by umanwizard** [comment]
I have changed `cloned` to `copies` in the test (and in the similar test for `from_timestamp_millis`), and moved the private method to the very bottom of the impl block.

File: src/naive/datetime/tests.rs [hunk 5]
```diff
@@ -17,20 +17,57 @@ fn test_datetime_from_timestamp_millis() {
...
+    let invalid = [i64::MAX, i64::MIN];
+
+    for timestamp_micros in invalid.iter().copied() {
+        let naive_datetime = NaiveDateTime::from_timestamp_micros(timestamp_micros);
+        assert!(naive_datetime.is_none());
+    }
+
+    // Test that the result of `from_timestamp_micros` compares equal to
+    // that of `from_timestamp_opt`.
+    let secs_test = [0, 1, 2, 1000, 1234, 12345678, -1, -2, -1000, -12345678];
```
Comment (esheppa) [inline src/naive/datetime/tests.rs:62]: Thanks for adding this here as well!

**Review by esheppa — APPROVED** [review]
Looks great!

**Pull request merged**